File: app/Models/Tenant/Announcement.php
<?php

namespace App\Models\Tenant;

use Illuminate\Database\Eloquent\Builder;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;
use Illuminate\Database\Eloquent\SoftDeletes;

class Announcement extends Model
{
    use SoftDeletes;

    protected $connection = 'society';

    public const STATUSES = ['draft', 'published', 'archived'];
    public const CATEGORIES = ['general', 'maintenance', 'event', 'urgent', 'other'];

    protected $fillable = [
        'title',
        'body',
        'category',
        'is_pinned',
        'status',
        'published_at',
        'expires_at',
        'posted_by_user_id',
    ];

    protected $casts = [
        'is_pinned' => 'boolean',
        'published_at' => 'datetime',
        'expires_at' => 'datetime',
    ];

    public function postedBy(): BelongsTo
    {
        return $this->belongsTo(User::class, 'posted_by_user_id');
    }

    /**
     * Scopes
     */
    public function scopeVisible(Builder $query): Builder
    {
        return $query->where('status', 'published')
            ->where(function ($q) {
                $q->whereNull('expires_at')->orWhere('expires_at', '>', now());
            });
    }

    public function scopePinnedFirst(Builder $query): Builder
    {
        return $query->orderByDesc('is_pinned')->orderByDesc('published_at');
    }

    public function isExpired(): bool
    {
        return $this->expires_at !== null && $this->expires_at->isPast();
    }
}

File: mobile/lib/features/announcements/data/announcement.dart
class Announcement {
  Announcement({
    required this.id,
    required this.title,
    required this.body,
    required this.category,
    required this.isPinned,
    this.postedBy,
    this.publishedAt,
    this.expiresAt,
  });

  factory Announcement.fromJson(Map<String, dynamic> json) {
    return Announcement(
      id: json['id'] as int,
      title: json['title'] as String,
      body: json['body'] as String,
      category: json['category'] as String,
      isPinned: json['is_pinned'] as bool? ?? false,
      postedBy: json['posted_by'] as String?,
      publishedAt: json['published_at'] as String?,
      expiresAt: json['expires_at'] as String?,
    );
  }

  final int id;
  final String title;
  final String body;
  final String category;
  final bool isPinned;
  final String? postedBy;
  final String? publishedAt;
  final String? expiresAt;
}
